x11-packages/xfce4-battery-plugin/libacpi.c

```diff
@@ -0,0 +1,130 @@
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+#ifndef __libacpi_c__
+#define __libacpi_c__
+#endif
+
+
+#include <stdlib.h>
+#include <stdio.h>
+#include "libacpi.h"
+
+
+char termux_bat_buffer[1024];
+
+int
+check_acpi(void)
+{
+    batt_count = 1;
+    return 0;
+}
+
+int read_termux_battery_status() {
+    FILE* fp;
+    size_t loc = 0;
+    size_t length = 0;
+    if ((fp = popen("termux-battery-status", "r")) == NULL) return 0;
```

Comment:
`termux-battery-status` may hang. It is better to kill it after a specified time if it doesn't output anything.

Reply:
I found that each `termux-battery-status` call is instantaneously cpu intensive. Would it be a better idea to rewrite to launch a long standing JVM instance for monitoring battery level? That should also resolve the time lag issue.

Reply:
How did you measure it? Are you using a termux-app version with [termux-am-socket](https://github.com/termux/termux-am-socket)? Check if `$TERMUX_APP__AM_SOCKET_SERVER_ENABLED` is exported and set to true.

Reply:
I used #20393. can be observed running `termux-battery-status`. or `top -d 0.25` shows similar result, but `top -d 0.5` does not. 

`$TERMUX_APP__AM_SOCKET_SERVER_ENABLED` is not set, where to do that? Or just `TERMUX_APP__AM_SOCKET_SERVER_ENABLED=true termux-battery-status`? Or somewhere else?

Reply:
Ah, it requires 0.119.0-beta1 from github or fdroid, and a recent termux-api app version as well. termux-app should then automatically export that var and use termux-am-socket if it exists

Reply:
I tried to launch JVM instance with `app_process`. A fake context can use the BatteryManager`, but the `BroadcastReceiver` never receives any incoming intent from the system. Is that normal and should I abandon the plan to use JVM callback? The code snippet is below FYR: 

```java
    public static void main(String[] args) throws InterruptedException {
        // adatped from scrcpy: 
        // https://github.com/Genymobile/scrcpy/blob/v2.4/server/src/main/java/com/genymobile/scrcpy/Workarounds.java
        // https://github.com/Genymobile/scrcpy/blob/v2.4/server/src/main/java/com/genymobile/scrcpy/FakeContext.java
        Context context;
        Looper.prepareMainLooper();
        try {
            Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
            Constructor<?> activityThreadConstructor = activityThreadClass.getDeclaredConstructor();
            activityThreadConstructor.setAccessible(true);
            Object activityThread = activityThreadConstructor.newInstance();

            Method getSystemContextMethod = activityThreadClass.getDeclaredMethod("getSystemContext");
            context = (Context) getSystemContextMethod.invoke(activityThread);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        BatteryManager batteryManager = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        System.out.println("BATTERY_PROPERTY_STATUS: " + batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_STATUS));

        context.registerReceiver(new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                int mBatteryStatus = intent.getIntExtra(BatteryManager.EXTRA_STATUS, BatteryManager.BATTERY_STATUS_UNKNOWN);
                System.out.println("mBatteryStatus = " + mBatteryStatus);
            }
        }, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));

        for (int i = 0; i< 10000; i++)
            Thread.sleep(100);
    }
```

Reply:
I don't know :) Maybe @agnostic-apollo or @tareksander have a better idea of the java side of things here

Reply:
https://github.com/termux/termux-x11/blob/360b6042911ad704cde1e7f40d2986695e654b32/app/src/main/java/com/termux/x11/CmdEntryPoint.java#L183-L211
You can go on using reflecion or you can use some stubs like [this](https://github.com/termux/termux-x11/tree/master/shell-loader/stub). Make sure you mention your stubs sources in [settings.gradle](https://github.com/termux/termux-x11/blob/360b6042911ad704cde1e7f40d2986695e654b32/settings.gradle#L1) and [link it to your app with compileOnly](https://github.com/termux/termux-x11/blob/360b6042911ad704cde1e7f40d2986695e654b32/app/build.gradle#L52)

Reply:
But AFAIK you can not register receiver without real context created by Android framework. Android has some code tracking processes created by system server and in the case if your process is not in the list it will not be allowed to register receiver or do a lot of other stuff.

Reply:
But you always can create a background thread which will poll battery status every 5 sec.

Reply:
About Thread.sleep. It should not work. Binder can not receive any incoming data en the case you are running Thread.sleep. It should work in the only case: 1. You have thread where you call `Looper.prepareMainLooper();` and `Looper.loop()`. And only this thread will receive remote connections so you should not run long-running methods there.

Reply:
I think in the longer term a redesign of Termux:API is planned, supporting use cases like longer lasting connections, so you don't have to pay the startup price each time.

Where is the CPU usage coming from? In case it's from `am`, then the am socket server implemented in 0.119.0-beta1 should help. You could try this out in an emulator if don't want to install the experimental version on your device.

Reply:
`termux-am-socket` is not available in `0.119.0-beta.1` as breaking changes were pending, will be available in next main beta. It's currently only available in github  action builds of `master` branch.

Will look into other comments above later.

Reply:
Thanks a lot, dev members! If `BroadcastReceiver` is not feasible, I will choose between `termux-battery-status` (requires `Termux:API`+`termux-api` and implementing timeout reading output, but there is official support) and `app_process` with `BatteryManager`. 

For the latter `app_process` approach, my only question left is perhaps how can JVM instance communicate back to the native code process. I suppose `app_process` creates a new Java process, so probably callback or JNI won't work and it will need IPC. Am I right and any suggestions?

Reply:
Will it be a plan or even posisble to use a real Android context in Termux Java? What hinders it, Android package policy?
I am intereted because: 
* currently adding Android features to Termux requires changes to Termux:App or Temrux:API which can be very slow to make it official. 
* Termux:API seems to mostly provide command line interface for terminal users, but not for native program in termux package or perhaps need `popen()` everytime. 
* still, the `app_process` approach comes with quite many limitations. 

It is a bit off-topic to this PR, would it be feasible to implement below features with `app_process`:  
* any `startActivityForResult()` equivalence? I know there is some `startActivity()` equivalence in `TermuxAm` but how about `ActivityForResult()`?
* any way to open a WebView? It is indeed a highly demanded feature (mentioned many times in reddit). I know about `Termux:GUI` but looks like a overkill for many users need who just want to beatifully show their locally hosted webpage, e.g. running a command like `termux-webview http://localhost:8888`.

Reply:
> so probably callback or JNI won't work and it will need IPC. Am I right and any suggestions?

JNI does not need IPC or something to work. It is a feature of JVM. Termux-x11 works with JNI without any issues.

Reply:
> real Android context in Termux Java?

It is only possible in the case if termux's service will instantiate new JVM with custom APK loading. Probably agnostic-apollo can implement that, but this will make coming back to google play impossible.

> any startActivityForResult() equivalence

What exactly you need? Your end goal. In the case of app_process started process it is not possible, but in the case if you want establish connection with your app you can use some Binder+broadcast workaround.

> any startActivityForResult() equivalence

You can open webpage in chrome or other browser with `termux-open` or `am start`.

Reply:
Context usage will be supported in future with use of reflection APIs, like how Tasker app allows running random java code with `Java Function` action. `startActivityForResult()` is not something too difficult and can maybe even be added to `termux-am-socket`, but maybe in Termux:API redesign where service will be used to keep asynchronous connections. WebView support has already been implemented for Media viewer APIs, but they aren't complete yet, and will be provided in future Termux app versions.